tests/python/pants_test/engine/scheduler_test_base.py

```diff
@@ -97,3 +100,120 @@ def execute_raising_throw(self, scheduler, product, subject):
     self.assertTrue(type(resulting_value) is Throw)
 
     raise resulting_value.exc
+
+
+class DeclarativeTaskTestBase(SchedulerTestBase):
+  """???/experimental blah, makes things declarative, whatever"""
+
+  @classproperty
+  def dist_specs(cls):
+    """
+    This is an informally-specified nested dict -- see ../test_ctypes.py for an example. Special
+    keys are 'key' (used to index into `self.target_dict`) and 'filemap' (creates files at the
+    specified relative paths). The rest of the keys are fed into `self.make_target()`. An
+    `OrderedDict` of 2-tuples may be used if targets need to be created in a specific order (e.g. if
+    they have dependencies on each other).
+    """
+    raise NotImplementedError('dist_specs must be implemented!')
+
+  @classproperty
+  def run_before_task_types(cls):
+    """
+    By default, we just use a `BuildLocalPythonDistributions` task. When testing with C/C++ targets,
+    we want to compile and link them as well to get the resulting dist to build, so we add those
+    task types here and execute them beforehand.
+    """
+    return []
+
+  @classproperty
+  def run_after_task_types(cls):
+    """Tasks to run after local dists are built, similar to `run_before_task_types`."""
+    return []
+
+  def populate_target_dict(self):
+    self.target_dict = {}
+
+    # Create a target from each specification and insert it into `self.target_dict`.
+    for target_spec, target_kwargs in self.dist_specs.items():
+      unprocessed_kwargs = target_kwargs.copy()
+
+      target_base = Address.parse(target_spec).spec_path
+
+      # Populate the target's owned files from the specification.
+      filemap = unprocessed_kwargs.pop('filemap', {})
+      for rel_path, content in filemap.items():
+        buildroot_path = os.path.join(target_base, rel_path)
+        self.create_file(buildroot_path, content)
```

Comment:
I think that this means that this needs to be mixed into something that extends `TestBase` anyway... so probably best to use the scheduler from `TestBase`.

Reply:
Was able to remove all scheduler nonsense! Noted that it needs to be mixed into something subclassing `TaskTestBase`!